packages/flutter/lib/src/cupertino/refresh.dart

```diff
@@ -0,0 +1,520 @@
+      case RefreshIndicatorMode.inactive:
+        if (lastIndicatorExtent <= 0) {
+          return RefreshIndicatorMode.inactive;
+        } else {
+          nextState = RefreshIndicatorMode.drag;
+        }
+        continue drag;
+      drag:
+      case RefreshIndicatorMode.drag:
+        if (lastIndicatorExtent == 0) {
+          return RefreshIndicatorMode.inactive;
+        } else if (lastIndicatorExtent < widget.refreshTriggerPullDistance) {
+          return RefreshIndicatorMode.drag;
+        } else {
+          if (widget.onRefresh != null) {
+            HapticFeedback.mediumImpact();
+            // Call onRefresh after this frame finished since the function is
+            // user supplied and we're always here in the middle of a the sliver's
+            // performLayout.
+            SchedulerBinding.instance.addPostFrameCallback((Duration timestamp) {
+              refreshTask = widget.onRefresh()..then((_) {
+                if (mounted) {
+                  setState(() => refreshTask = null);
+                  // Trigger one more transition because by this time, BoxConstraint's
+                  // maxHeight might already be resting at 0 in which case no
+                  // calls to [transitionNextState] will occur anymore and the
+                  // state may be stuck in a non-inactive state.
+                  refreshState = transitionNextState();
+                }
+              });
+              setState(() => hasSliverLayoutExtent = true);
+            });
+          }
+          return RefreshIndicatorMode.armed;
+        }
+        // Don't continue here. We can never possibly call onRefresh and
+        // progress to the next state in one [computeNextState] call.
+        break;
+      case RefreshIndicatorMode.armed:
+        if (refreshState == RefreshIndicatorMode.armed && refreshTask == null) {
+          goToDone();
+          continue done;
+        }
+
+        if (lastIndicatorExtent > widget.refreshIndicatorExtent) {
+          return RefreshIndicatorMode.armed;
+        } else {
+          nextState = RefreshIndicatorMode.refresh;
+        }
+        continue refresh;
+      refresh:
+      case RefreshIndicatorMode.refresh:
+        if (refreshTask != null) {
+          return RefreshIndicatorMode.refresh;
+        } else {
+          goToDone();
+        }
+        continue done;
+      done:
+      case RefreshIndicatorMode.done:
+        // Let the transition back to inactive trigger before strictly going
+        // to 0.0 since the last bit of the animation can take some time and
+        // can feel sluggish if not going all the way back to 0.0 prevented
+        // a subsequent pull-to-refresh from starting.
+        if (lastIndicatorExtent >
+            widget.refreshTriggerPullDistance * _kInactiveResetOverscrollFraction) {
+          return RefreshIndicatorMode.done;
+        } else {
+          nextState = RefreshIndicatorMode.inactive;
+        }
+        break;
+      default:
```

Comment:
Don't put a default here.  You've handled all the cases, so it shouldn't be necessary, and will highlight a missing case if more modes are added in the future.

Reply:
good idea, done